content/advent-2019/directional-channels.md

```diff
@@ -0,0 +1,193 @@
++++
+title= "Directional Channels in Go"
+date = "2019-12-03T00:00:00+00:00"
+series = ["Advent 2019"]
+author = ["Andy Walker"]
+linktitle = "flag"
++++
+
+Go's [channels](https://tour.golang.org/concurrency/2) provide a primitive for typed, synchronous message passing. Combined with [goroutines](https://tour.golang.org/concurrency/1), they form the backbone of Go's [CSP](https://en.wikipedia.org/wiki/Communicating_sequential_processes)-inspired concurrency model. They're simple and expressive, but they can be difficult to use properly, especially if you need to control who can read from them or write to them.
+
+## The Problem With Bidirectional Channels
+
+Channels are normally declared with the `chan` keyword, followed by the **ElementType**, which represents the type of values passed on that channel. Together, these form the composite type for any value, which you can inspect with `%T`.
+
+```go
+var stringChan chan string
+fmt.Printf("%T\n", stringChan) // "chan string"
+```
+[playground](https://play.golang.org/p/F58BWz2HJEZ)
+
+This is the declaration format most people first encounter when working with channels. But any channel created in this way will be bidirectional as the default behavior. This means that anyone who has access to a channel value can read from it *and* write to it. This can cause problems in a concurrent environment, and many a Go programmer has torn hair from their heads trying to debug a `panic: send on a closed channel`.
+
+The common wisdom is that only the sender should close a channel, and this makes sense. Only the sender can know when there's no more data to send, and it's the receiver's responsibility to watch for the close, or ideally, to simply `range` over the channel, exiting the loop naturally when it's done. If this order is upset, it's generally a sign something very wrong is going on, hence the panic. But if anyone can perform any action on a channel, including calling `close()`, how can you reel this in?
```

Comment:
This turned into a bit of a run-on.  

"Common wisdom is that the sender is responsible for closing the channel.  This makes sense, since only the sender can know when there is no more data to send.  It's the receiver's responsibility to watch for and handle the channel closing.  Go's `range` construct will exit the associated `for` loop when the channel is closed, making it an ideal way to structure the receiver.  When a channel is closed incorrectly, those panics signal that something very wrong is happening.  So, how do we make sure that the correct goroutine calls `close()` when all of them are able?